### PR TITLE
142 Resolver setup

### LINKED ISSUE #142: Name resolution
**As a** researcher
**I want to** be able to able to find a substance using a name or casrn
**So that** I can find the chemical I am interested in

`GET https://api.chemreg.epa.gov/search?identifer=$$$$$$$`

Acceptance Criteria:
- An endpoint exists that allows me to search for an identifier.  
- that endpoint being a different and complicated function spanning multiple resources in our CRUD API should exist as a separate microservice
- as a separate microservice and following the concept of encapsulation, the resolution app should have a separate data source that is not coupled at the database level to our CRUD database
- If the searched identifier matches a substance.preferred_name, substance.display_name, substance.casrn, a list of the dtxsids for those substances are returned.
- Separate repo and docker file
- New datastore

### PR DESCRIPTION
**Pull request by mwfrost**
Closes: https://github.com/Chemical-Curation/chemcurator_django/issues/142

- [x] `docker-compose` file with services `db`, `web`, and `pg-bouncer`
- [x] PostgreSQL database in Docker with basic user auth schema and Substances
- [x] Flask REST API with endpoints for CRUD on the `Substance` model
  - [x] Working GET requests on `/substances`
  - [x] Working GET requests on `/resolver`
  - [x] Working ~PUT~ PATCH requests on `/substances`
    - [x] Returning serialized version of new instance in response
- [x] Basic search functionality on `Substance` attributes
- [x] pytest working, tests for Substance CRUD via API
  - [x] pytest successfully loading `psycopg2` into the test environment
- [x] Preliminary Swagger documentation of API
- [x] Preliminary seed data for testing (only works via API POST requests)
- [x] developer setup documentation

### CODE REVIEW

**Comment by mwfrost** [comment]
I have not gotten the API to successfully POST the JSON data into the `substance.identifiers` field,, so rather than use a Postman request, I've got this script. 

```
COPY public.substances (id, identifiers) FROM stdin;
DTXSID302000003	{"casrn": "1050-79-9", "casalts": [{"casalt": "0001050799", "weight": 0.5}, {"casalt": "1050799", "weight": 0.5}], "inchikey": "AGAHNABIDCTLHW-UHFFFAOYSA-N", "synonyms": [{"weight": 0.75, "synonym": "Meperon"}, {"weight": 0.5, "synonym": "Methylperidol"}], "preferred_name": "Moperone"}
DTXSID302000004	{"casrn": "7722-84-1", "casalts": [{"casalt": "0007722841", "weight": 0.5}, {"casalt": "7722841", "weight": 0.5}], "inchikey": "MHAJPDPJQMAIIY-UHFFFAOYSA-N", "synonyms": [{"weight": 0.75, "synonym": "Hydrogen peroxide [USP]"}, {"weight": 0.5, "synonym": "Wasserstoffperoxid"}], "preferred_name": "Hydrogen peroxide"}
\.

```

**Comment by mwfrost** [comment]
Here is the repo for the cookiecutter I used:
https://github.com/karec/cookiecutter-flask-restful

**Comment by michael-barbour** [comment]
@mwfrost I built ticket `142-setup-jsonapi-rewrites` as a playground for large changes regarding `flask-rest-jsonapi` package.  `142-setup-jsonapi-rewrites` should be merged into this branch before any more changes are made here.

It's currently torn between the `flask-restful` package from the cookiecutter you used as a base, and a full jsonapi implementation.  The jsonapi package [does not utilize swagger](https://github.com/miLibris/flask-rest-jsonapi/issues/12) (stop me if you've heard this one before) and we will need to write the swagger docs by hand as far as I can tell.

Tests are all running, but there are none in place for search.  I also couldn't get search working on your branch but hadn't looked at it extensively.

Docker builds and runs but with the pgbouncer pod commented out.  Scalability seems a far ways out.

Developer setup [wiki](https://github.com/Chemical-Curation/resolver/wiki/Developer-Setup) was updated.  It may need more work on the docker portion but should hopefully guide through the trickier parts of setup.

**Comment by mwfrost** [comment]
To test this branch, load seed data with the three POST requests in `/tests/Resolver.postman_collection.json`. The other requests in that collection demonstrate the search functionality.

File: resolver/manage.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    db.session.add(s1)
+    s2 = Substance(
+        id="DTXCID302000004",
+        identifiers="{'preferred_name':'Hydrogen peroxide','casrn':'7722-84-1',"
+        "'inchikey': 'MHAJPDPJQMAIIY-UHFFFAOYSA-N', "
+        "'casalts':[{'casalt':'0007722841','weight':0.5},{'casalt':'7722841','weight':0.5}],"
+        "'synonyms': [{'synonym': 'Hydrogen peroxide [USP]','weight': 0.75},"
+        "{'synonym': 'Wasserstoffperoxid','weight': 0.5}]}",
+    )
+
```
Comment (michael-barbour) [inline resolver/manage.py:64]: This loads as a string not as json.  The posts work so we will be leaning on the api CRUD operations for now.

**Review by michael-barbour — APPROVED** [review]
This looks good.

File: resolver/manage.py (hunk 1 above)
Comment (mwfrost) [reply resolver/manage.py:64]: Correct. I have not gotten any seed data to load properly through the ORM.

Comment (debboutr) [reply resolver/manage.py:64]: has this been tried...?
```
In [1]: import json

In [2]: json.loads('{"preferred_name":"Hydrogen peroxide","casrn":"7722-84-1","inchikey": "MHAJPDPJQMAIIY-UHFFFAOYSA-N","casalts":[{"casalt":"0007722841","weight":0.5},{
   ...: "casalt":"7722841","weight":0.5}],"synonyms": [{"synonym": "Hydrogen peroxide [USP]","weight": 0.75},{"synonym": "Wasserstoffperoxid","weight": 0.5}]}')
Out[2]: 
{u'casalts': [{u'casalt': u'0007722841', u'weight': 0.5},
  {u'casalt': u'7722841', u'weight': 0.5}],
 u'casrn': u'7722-84-1',
 u'inchikey': u'MHAJPDPJQMAIIY-UHFFFAOYSA-N',
 u'preferred_name': u'Hydrogen peroxide',
 u'synonyms': [{u'synonym': u'Hydrogen peroxide [USP]', u'weight': 0.75},
  {u'synonym': u'Wasserstoffperoxid', u'weight': 0.5}]}
```

Comment (debboutr) [reply resolver/manage.py:64]: this format is working for me...
```
    db.session.add(s1)
    s2 = Substance(
        id="DTXCID302000004",
        identifiers={"preferred_name":"Hydrogen peroxide","casrn":"7722-84-1",
        "inchikey": "MHAJPDPJQMAIIY-UHFFFAOYSA-N",
        "casalts":[{"casalt":"0007722841","weight":0.5},{"casalt":"7722841","weight":0.5}],
        "synonyms": [{"synonym": "Hydrogen peroxide [USP]","weight": 0.75},
        {"synonym": "Wasserstoffperoxid","weight": 0.5}]},
    )

```

**Pull request merged**